ddtrace/writer.py

```diff
@@ -105,51 +106,99 @@ def _on_shutdown(self):
-                    time.sleep(0.05)
+                # Block until all items have been removed from the queue
+                self._trace_queue.join()
+
+    def _filter_trace(self, trace):
+        if not trace:
+            return None
+
+        if not self._filters:
+            return trace
+
+        for f in self._filters:
+            trace = f.process_trace(trace)
+            if trace is None:
+                log.debug('filter %r filtered out trace %r', f, trace)
+                return None
+
+        return trace
 
     def _target(self):
-        traces_response = None
+        next_flush = time.time() + 1
+        payload = self.api.new_payload()
 
         while True:
-            traces = self._trace_queue.pop()
-            if traces:
-                # Before sending the traces, make them go through the
-                # filters
-                try:
-                    traces = self._apply_filters(traces)
-                except Exception as err:
-                    log.error('error while filtering traces:{0}'.format(err))
-            if traces:
-                # If we have data, let's try to send it.
-                try:
-                    traces_response = self.api.send_traces(traces)
-                except Exception as err:
-                    log.error('cannot send spans to {1}:{2}: {0}'.format(err, self.api.hostname, self.api.port))
-
-            if self._trace_queue.closed() and self._trace_queue.size() == 0:
-                # no traces and the queue is closed. our work is done
+            # DEV: Returns `None` or a Trace
+            trace = self._trace_queue.get()
+
+            # Filter the trace
+            # DEV: We set `trace` to `None` on an error because we want to continue the loop
+            #      since this loop needs to execute in order to flush a payload if enough time
+            #      has passed since the last flush.
+            try:
+                # DEV: Returns the trace or `None` if it was filtered
+                trace = self._filter_trace(trace)
+            except Exception as err:
+                log.error('Error raised while trying to filter trace %r: %s', trace, err)
+                trace = None
+
+            if trace:
+                log.debug('Popped trace with %s spans', len(trace))
+                payload.add_trace(trace)
+
+            now = time.time()
+            # If the payload is large enough or enough time has passed, then flush
+            if payload.full or now >= next_flush:
+                if not payload.empty:
+                    log.debug('Attempting to flush %r', payload)
+                    self._flush(payload)
+
+                    # Get a fresh payload
+                    # DEV: Payloads are not reusable
+                    payload = self.api.new_payload()
+                next_flush = now + 1
+
+            # Wait up until the next flush for another trace to be added
+            self._trace_queue.wait(max(0, next_flush - now))
```

Comment:
I don't know how precise you want to be, but the flushing could take some time, so your `now` might be a little be outdated at that point.